semgrep-core/src/analyzing/Constant_propagation.ml

```diff
@@ -694,4 +694,14 @@ let propagate_dataflow lang ast =
       |> Visit_function_defs.visit (fun _ent fdef ->
              let inputs, xs = AST_to_IL.function_definition lang fdef in
              let flow = CFG_build.cfg_of_stmts xs in
-             propagate_dataflow_one_function lang inputs flow)
+             propagate_dataflow_one_function lang inputs flow);
+
+      (* We consider the top-level function the interior of a degenerate function,
```

Comment:
This may have some problems, I presume you'll get a match for pattern `f(42)` here:

```python
x = 42

def update_x():
  global x
  x = 0

update_x()
f(x)
```

When you should not get any. This could get confusing, especially with symbolic-propagation which works more often than constant propagation. That's why at the time, I proposed that we restricted this to variables that we can guarantee (to some extent) that they are not being modified inside functions. It's less problematic for taint I think because it's a may-analysis, although to be fair we should do the same there too.

It's fine for now but just to keep it in mind.